pkg/storage/concurrency/concurrency_control.go

```diff
@@ -0,0 +1,355 @@
+// and no conflicting lockWaitQueues that it is not already the head of, the
+// request can proceed to evaluate. During evaluation, the request may insert or
+// remove locks from the lockTable for its own transaction. This is performed
+// transparently by a lockAwareBatch/lockAwareReadWriter. The request may also
+// need to consult its own locks to properly interpret the corresponding intents
+// in the MVCC keyspace. This is performed transparently by a lockAwareIter.
+//
+// When the request completes, it exits any lockWaitQueues that it was at the
+// head of and releases its latches. However, any locks that it inserted into
+// the lockTable remain.
+type Manager interface {
+	requestSequencer
+	contentionHandler
+	transactionManager
+	rangeUpdateListener
+	metricExporter
+}
+
+// requestSequencer is concerned with the sequencing of concurrent requests.
+type requestSequencer interface {
+	// SequenceReq acquires latches, checks for locks, and queues behind and/or
+	// pushes other transactions to resolve any conflicts. Once sequenced, the
+	// request is guaranteed sufficient isolation for the duration of its
+	// evaluation, until its guard is released.
+	// NOTE: this last part will not be true until replicated locks are pulled
+	// into the concurrency manager.
+	//
+	// An optional existing request guard can be provided to SequenceReq. This
+	// allows the request's position in lock wait-queues to be retained across
+	// sequencing attempts. If provided, the guard should not be holding latches
+	// already. This typically means that the guard was acquired through a call
+	// to a contentionHandler method.
+	//
+	// If the method returns a non-nil request guard then the caller must ensure
+	// that the guard is eventually released by passing it to FinishReq.
+	//
+	// Alternatively, the concurrency manager may be able to serve the request
+	// directly, in which case it will return a Response for the request. If it
+	// does so, it will not return a request guard.
+	SequenceReq(context.Context, *Guard, Request) (*Guard, Response, *Error)
+	// FinishReq marks the request as complete, releasing any protection
+	// the request had against conflicting requests and allowing conflicting
+	// requests that are blocked on this one to proceed. The guard should not
+	// be used after being released.
+	FinishReq(*Guard)
+}
+
+// contentionHandler is concerned with handling contention-related errors. This
+// typically involves preparing the request to be queued upon a retry.
+type contentionHandler interface {
+	// HandleWriterIntentError consumes a WriteIntentError by informing the
+	// concurrency manager about the replicated write intent that was missing
+	// from its lock table. After doing so, it enqueues the request that hit the
+	// error in the lock's wait-queue (but does not wait) and releases the
+	// guard's latches. It returns an updated guard reflecting this change.
```

Comment:
When these methods return a new `*Guard`, does it mean that the original `*Guard` must no longer be used?

Reply:
Done.